clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -4434,6 +4453,118 @@ void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
   D->addAttr(::new (Context) SYCLIntelMaxGlobalWorkDimAttr(Context, CI, E));
 }
 
+void Sema::AddSYCLIntelMinWorkGroupsPerComputeUnitAttr(
+    Decl *D, const AttributeCommonInfo &CI, Expr *E) {
+  if (Context.getLangOpts().SYCLIsDevice &&
+      !Context.getTargetInfo().getTriple().isNVPTX()) {
+    Diag(E->getBeginLoc(), diag::warn_launch_bounds_is_cuda_specific)
+        << CI << E->getSourceRange();
+    return;
+  }
+  if (!E->isValueDependent()) {
+    uint32_t Val;
+    if (!checkUInt32Argument(*this, CI, E, Val, UINT_MAX /* Idx */,
+                             true /* StrictlyUnsigned */))
+      return;
+
+    // Validate that we have an integer constant expression and then store the
```

Comment:
Perhaps as a side note, `VerifyIntegerConstantExpression` has this nice property (I'm slightly out of my depths with clang, so forgive if I misname the things) that it rewrites the AST, so for each use of the attributes, it always follows the same pattern:
```
ConstantExpr{{.*}}'int'      
value: Int N                 
IntegerLiteral {{.*}} 'int' N
```

Reply:
Hm, that exactly how it used to be, I had it in a helper function used by both attributes: https://github.com/intel/llvm/pull/11192/commits/9431687c6bfe57ab8f06cdabff4273a73c0abf9f#diff-2a5bdb2d9f07f8d77de51d5403d349c22978141b6de6bd87fc5e449f5ed95becL4450-L4469

Reply:
Yea sorry for some confusion over this. I thought you would be able to just call `checkUInt32Argument` (without calling `VerifyIntegerConstantExpression`, but I am not so sure about this anymore. I've asked a colleague about this and will get back to you once he replies). In any case though, the issue with the code you first had and the current one now is with calling `getIntegerConstantExpr` or it's equivalent twice. `VerifyIntegerConstantExpression` should give you the value you need without having to call it twice right?

Reply:
@smanna12 please confirm but I do not think you need to do both checkUInt32Argument and VerifyIntegerConstantExpression.

Yes, both does not need to use together. You can just use the Value from VerifyIntegerConstantExpression and replicate this diagnostic from checkUInt32Argument

Reply:
```
 // This attribute requires a strictly positive value.
      if (ArgVal <= 0) {
        Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
            << CI << /*positive*/ 0;
        return std::nullopt;
      }
    }
```

Reply:
@jchlanda I spoke to Aaron who has implemented/worked with a lot of attribute code in community clang to get a better understanding of when to call `checkUInt32Argument` vs `VerifyIntegerConstantExpression`.  Apparently `checkUInt32Argument` was implemented when template instantiations of arguments were not really a thing. Since we're now doing more with template instantiations, it makes sense to catch the constant expression  (`checkUInt32Argument` discards it after evaluating it) and keep it in AST. So `VerifyIntegerConstantExpression` + required diagnostics as suggested above is the right way to handle this. I apologize for the confusion here. This area needs some cleanup in community clang.

Reply:
Sure, not a problem at all. I got rid of the extra overload to make `checkUInt32Argument` work, and encapsulated those check in a static helper function that is called instead of `checkUInt32Argument` (while keeping `VerifyIntegerConstantExpression`), it required some test changes, all commited here: acba95f8d561c66ec83fb33b56c027a2f7648365

Reply:
> ```
>  // This attribute requires a strictly positive value.
>       if (ArgVal <= 0) {
>         Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
>             << CI << /*positive*/ 0;
>         return std::nullopt;
>       }
>     }
> ```
> 
> > > I'm guessing there are other attributes which do similar things. @smanna will know better since she works with these a lot.
> 
> SYCLIntelBankBitsAttr is doing similar thing in void Sema::AddSYCLIntelBankBitsAttr()

Please stop using @smanna.

Reply:
> ```
>  // This attribute requires a strictly positive value.
>       if (ArgVal <= 0) {
>         Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
>             << CI << /*positive*/ 0;
>         return std::nullopt;
>       }
>     }
> ```
> 
> > > I'm guessing there are other attributes which do similar things. @smanna will know better since she works with these a lot.
> 
> SYCLIntelBankBitsAttr is doing similar thing in void Sema::AddSYCLIntelBankBitsAttr()

Please stop using @smanna.